lib/puppet/util/windows/registry.rb

```diff
@@ -30,51 +32,276 @@ def open(name, path, mode = KEY_READ | KEY64, &block)
       end
     end
 
-    def values(subkey)
-      values = {}
-      subkey.each_value do |name, type, data|
-        case type
-        when Win32::Registry::REG_MULTI_SZ
-          data.each { |str| force_encoding(str) }
-        when Win32::Registry::REG_SZ, Win32::Registry::REG_EXPAND_SZ
-          force_encoding(data)
+    def keys(key)
+      keys = {}
+      each_key(key) { |subkey, filetime| keys[subkey] = filetime }
+      keys
+    end
+
+    # subkey is String which contains name of subkey.
+    # wtime is last write time as FILETIME (64-bit integer). (see Registry.wtime2time)
+    def each_key(key, &block)
+      index = 0
+      subkey = nil
+
+      subkey_max_len, value_max_len = reg_query_info_key_max_lengths(key.hkey)
+
+      begin
+        subkey, filetime = reg_enum_key(key.hkey, index, subkey_max_len)
+        yield subkey, filetime
+        index += 1
+      end while !subkey.nil?
+
+      index
+    end
+
+    def values(key)
+      vals = {}
+      each_value(key) { |subkey, type, data| vals[subkey] = data }
+      vals
+    end
+
+    def each_value(key, &block)
+      index = 0
+      subkey = nil
+
+      subkey_max_len, value_max_len = reg_query_info_key_max_lengths(key.hkey)
+
+      begin
+        subkey, type, data = reg_enum_value(key.hkey, index, value_max_len)
+        yield subkey, type, data
+        index += 1
+      end while !subkey.nil?
+
+      index
+    end
+
+    private
+
+    def reg_enum_key(hkey, index, max_key_length = Win32::Registry::Constants::MAX_KEY_LENGTH)
+      subkey, filetime = nil, nil
+
+      FFI::MemoryPointer.new(:dword) do |subkey_length_ptr|
+        FFI::MemoryPointer.new(FFI::WIN32::FILETIME.size) do |filetime_ptr|
+          FFI::MemoryPointer.new(:wchar, max_key_length) do |subkey_ptr|
+            subkey_length_ptr.write_dword(max_key_length)
+
+            # RegEnumKeyEx cannot be called twice to properly size the buffer
+            result = RegEnumKeyExW(hkey, index,
+              subkey_ptr, subkey_length_ptr,
+              FFI::Pointer::NULL, FFI::Pointer::NULL,
+              FFI::Pointer::NULL, filetime_ptr)
+
+            break if result == ERROR_NO_MORE_ITEMS
```

Comment:
Does this break out of this method? or does it break out of the caller's loop in https://github.com/puppetlabs/puppet/pull/3540/files#diff-85725d2700dc7373579506c97d84a83eR60? In either case, it might be clearer to do:

<pre>
if result != ERROR_NO_MORE_ITEMS
  if result !=FFI::ERROR_SUCCESS
    raise ...
  end
  ...
end

[subkey, filetime]
</pre>